### PR TITLE
coinex - transfer and find leverage in fetch Market

### PR DESCRIPTION
**Pull request by pcriadoperez**
There is another wallet called financial, however there is no api call in the docs to allow transfers to it. I checked the call the frontend does but seemed to different for us to include here.

### CODE REVIEW

File: js/coinex.js [hunk 1]
```diff
@@ -246,10 +262,34 @@ module.exports = class coinex extends Exchange {
         //                 "trading_name": "WAVES",
         //                 "trading_decimal": 8
         //             }
+        //             ...
         //         }
         //     }
         //
-        const markets = this.safeValue (response, 'data', {});
+        //     privateGetMarginConfig
```
Comment (pcriadoperez) [inline js/coinex.js:269]: As this call is private and the first one is private I'm thinking it may be best to be behind some kind of option?

File: js/coinex.js [hunk 2]
```diff
@@ -990,6 +1031,100 @@ module.exports = class coinex extends Exchange {
...
+        let method = undefined;
+        if (fromAccount === 'margin' || toAccount === 'margin') {
+            method = 'privatePostMarginTransfer';
+            const symbol = this.safeString (params, 'symbol');
+            if (symbol === undefined) {
+                throw new ExchangeError ('You must define in params what market you are transferring. e.g. {symbol: ETH/BTC}');
+            }
+            const market = this.market (symbol);
+            request['market'] = market['id'];
+            const marginMarkets = await this.publicGetMarginMarket (params);
```
Comment (pcriadoperez) [inline js/coinex.js:1053]: At the beginning I thought we could avoid this call by asking to pass the account id in the params. However I think this is better as it feels more unified for the user to pass in the margin symbol in params and for us to find the accountId, as I've done this for another `transfer` function too.

This call at the beginning at the beginning I put into `fetchAccounts` so it could be in the cache with `this.loadAccounts` however it seemed a bit to hard to parse to current `fetchAccounts` structure.

File: js/coinex.js (hunk 1 above)
Comment (kroitor) [reply js/coinex.js:268]: We can check if the exchange instance has the necessary credentials and only issue that call, if those credentials are available, see how it's done here, for example: https://github.com/ccxt/ccxt/blob/master/js/binance.js#L1224

File: js/coinex.js (hunk 2 above)
Comment (kroitor) [reply js/coinex.js:1052]: > This call at the beginning at the beginning I put into `fetchAccounts` so it could be in the cache with `this.loadAccounts` however it seemed a bit to hard to parse to current `fetchAccounts` structure.

How does that response structure of fetchAccounts look like?

File: js/coinex.js [hunk 3]
```diff
@@ -990,6 +1030,100 @@ module.exports = class coinex extends Exchange {
...
+        const request = {
+            'coin_type': currency['id'],
+            'amount': this.currencyToPrecision (code, amount),
+        };
+        let method = undefined;
+        if (fromAccount === 'margin' || toAccount === 'margin') {
+            method = 'privatePostMarginTransfer';
+            const symbol = this.safeString (params, 'symbol');
+            if (symbol === undefined) {
+                throw new ExchangeError ('You must define in params what market you are transferring. e.g. {symbol: ETH/BTC}');
```
Comment (kroitor) [inline js/coinex.js:1048]: Let's use this uniform exception message format:

```JavaScript
throw new ArgumentsRequired (thid.id + ' transfer() requires a symbol parameter')
throw new ArgumentsRequired (thid.id + ' createOrder() requires a price argument')
throw new ArgumentsRequired (thid.id + ' fetchTransfers() requires an account_id parameter')
...
```

Also, let's make sure we don't forget to omit `symbol` from `params` before sending out the request.

Logically, we must support the exchange-specific parameter name `market` for the marketId as well as the unified parameter name `symbol` for unified market symbols.

1. Check if `market` is specified in `params`.
2. If `params['market']` is specified, then it contains the market id.
3. If it's not specified, check if `symbol` is specified.
4. If `params['symbol']` is specified, convert the unified symbol to the market id and omit the `symbol` from `params`, otherwise it will be sent out to the exchange (that does not support the `symbol` parameter) because the request is extended with `params` later upon sending out the request.

```JavaScript
if (fromAccount === 'margin' || toAccount === 'margin') {
    const marketId = this.safeString (params, 'market');
    // if not specified, check the symbol parameter as well
    // otherwise, if specified, it will be sent automatically
    // because the request is extended with params before sending it out
    if (marketId === undefined) {
        const symbol = this.safeString (params, 'symbol');
        if (symbol === undefined) {
            throw new ExchangeError (this.id + ' transfer() requires an exchange-specific market parameter or a unified symbol parameter');
        } else {
            params = this.omit (params, 'symbol');
            const market = this.market (symbol);
            request['market'] = market['id'];
            // ...
        }
    }
    // ...
```

File: js/coinex.js [hunk 4]
```diff
@@ -990,6 +1030,100 @@ module.exports = class coinex extends Exchange {
...
+            }
+            const market = this.market (symbol);
+            request['market'] = market['id'];
+            const marginMarkets = await this.publicGetMarginMarket (params);
```
Comment (kroitor) [inline js/coinex.js:1052]: I'm not sure if we should do this `publicGetMarginMarket()` request on each call to `transfer()`. The rule of thumb is to have one HTTP request per unified method, with very few exceptions. In this case, I think  `publicGetMarginMarket()` belongs to `fetchMarkets()`, where we can issue more than one request, loading and caching them on the fly. So, I suggest that we move this call to `fetchMarkets` once and reuse this info on consequent calls to `transfer()`.

For example: 

- https://github.com/ccxt/ccxt/blob/master/js/gateio.js#L640

File: js/coinex.js (hunk 2 above)
Comment (pcriadoperez) [reply js/coinex.js:1052]: So fetchAccounts should return something like: 
```
[
    {
        id: "s32kj302lasli3930",
        type: "main",
        currency: "USDT",
        info: { ... }
    },
```

Coinex when calling `publicGetMarginMarket` returns:
```
{
   "BTCUSDT": 1,    # BTCUSDT:market, 1:account ID
   "BCHUSDT": 2,
   ......
 },
 ```
 And it also has subAccounts that can be found with `privateGetSubAccountBalance` that returns:
 ```
 data: {
    newsubaccount: {
      ZIL: { available: '0', frozen: '0' },
      BAKE: { available: '0', frozen: '0' },
      ...
      },
      ...
  }
```
Also I thought maybe it makes sense during fetchMarkets to get the accountId of the margin symbols, and store it in the market? And keep `fetchAccounts` for the subaccounts in this case?

File: js/coinex.js (hunk 4 above)
Comment (pcriadoperez) [reply js/coinex.js:1052]: Sounds good! I thought the same, I just wasn't sure if it was ok to add `accountId` to the market object. I'll make the change

File: js/coinex.js (hunk 2 above)
Comment (pcriadoperez) [reply js/coinex.js:1052]: After reading the comment below I'll move `publicGetMarginMarket` to `fetchMarkets`

**Review by carlotestor — COMMENTED** [review]
Thanks for this, and apologies it sat open so long — this is a 2022 PR and the tree has moved a long way underneath it. Posting a status review rather than a code review, because the branch can no longer be rebased as written.

Merge gate: 🔴

**Category: SUPERSEDED (transfer) / STALE (leverage in fetchMarkets)** — the two halves have different fates, so they are graded separately below.

## 1. What conflicts

The PR's only file is `js/coinex.js` (+125/−39). That file no longer exists on master:

```
$ git cat-file -e origin/master:js/coinex.js
fatal: path 'js/coinex.js' does not exist in 'origin/master'
$ git ls-tree origin/master ts/src/coinex.ts
100644 blob c76231626a1cde93a13bfa2b6bbac549fa253198	ts/src/coinex.ts
```

So this is a **modify/delete conflict with no rebase target**. The migration happened in two steps, both after this PR opened (2022-04-08):

- `41ac5112f4c` *"rename to ts"* (2022-07-01) — `js/coinex.js` → `ts/coinex.ts`
- `82725d8a562` *"move inside src"* (2022-10-15) — `ts/coinex.ts` → `ts/src/coinex.ts`

Since then `ts/src/` is the single source of truth and `js/` is generated output only. Any change here has to be authored in `ts/src/coinex.ts`; there is no mechanical way to replay this diff.

## 2. Is it already landed?

### (a) `transfer` — SUPERSEDED, and by your own work

The transfer half landed on master long ago, and then was rewritten onto CoinEx's v2 API:

- `87522b94c1a` *"Coinex transfer"* (2022-05-03) — added `has.transfer`, `transfer()` and `parseTransfer()`
- `67cfadce76e` *"Coinex transfer"* (2022-05-30) — added the `privatePostMarginTransfer` margin branch this PR builds
- `97d29a571f8` *"coinex: transfer v2 (#22511)"* — replaced the whole method with the unified v2 endpoint
- `2b20cac86fe` *"fix(coinex): transfer account id (#24165)"* — later `accountsById`→`accountsByType` fix

Master today (`ts/src/coinex.ts`) has `'transfer': true` in `has`, an `override async transfer (...)` posting to `v2PrivatePostAssetsTransfer`, plus `parseTransfer` and `parseTransferStatus`. The v2 endpoint takes `from_account_type` / `to_account_type` directly, so the whole `accountId`-per-market machinery this PR introduces (`request['from_account'] = accountId`, `privatePostContractBalanceTransfer`, `privatePostSubAccountTransfer`, method-name dispatch via `this[method]`) has no counterpart on master — `grep -n accountId ts/src/coinex.ts` returns nothing. The margin `symbol` requirement you introduced *did* survive in spirit: master throws `ArgumentsRequired` when either side is `margin` and no `symbol` is given.

Nothing to salvage here.

### (b) leverage discovery in `fetchMarkets` — NOT landed for spot/margin

This half is genuinely still open, but only partially and not in the shape proposed.

- `42bc15fbd5e` *"coinex: fetchMarkets v2 (#22180)"* rebuilt `fetchMarkets` as `Promise.all([fetchSpotMarkets, fetchContractMarkets])`. `fetchContractMarkets` **does** derive leverage bounds, straight from the v2 payload's `leverage` array (`ts/src/coinex.ts` ~L950, L994-997): `'min': this.safeNumber (leverages, 0)`, `'max': this.safeNumber (leverages, leveragesLength - 1)` — no extra request, no credentials.
- `fetchSpotMarkets` still emits `'leverage': { 'min': undefined, 'max': undefined }` (~L898-901). So spot/margin markets carry no leverage bounds, which is the gap this PR aimed at.
- The `margin` flag part *is* covered: `9c11f72aca4` *"fix(coinex): add margin property (#26350)"* set `'margin': this.safeBool (market, 'is_margin_available')` from the v2 spot payload — so master no longer needs a `margin/config` round-trip to know whether a spot market is marginable, which was the other thing your `margin: leverage !== undefined` derivation bought.

## 3. Is anything still wanted?

Only the narrow residue: **spot/margin markets have empty `limits.leverage`**. If you still care about that, it is worth a fresh, small PR against `ts/src/coinex.ts` — but not this implementation, for reasons that are about master's current design rather than about the code you wrote:

- It targets the **v1** endpoints (`publicGetMarketInfo`, `publicGetMarginMarket`, `privateGetMarginConfig`). Master's `fetchMarkets` is entirely v2; `margin/market` and `margin/config` still exist in the v1 `api` map but are no longer on the market-loading path.
- `const getMarginData = this.checkRequiredCredentials (false)` makes `fetchMarkets` **credentialed-dependent** — the same exchange would return different market metadata with and without keys. That is a hard sell now: static request tests and `loadMarkets()` are expected to be key-independent.
- Three parallel requests on every `loadMarkets()` (four once contracts are added) is a meaningful cost increase for a field most callers do not read.

If CoinEx's v2 spot payload carries margin leverage anywhere, reading it there — the way `fetchContractMarkets` reads `leverage` — would be the clean shape and a much smaller diff.

## Verdict

| Half | Category | Evidence |
|---|---|---|
| `transfer` | 🔴 SUPERSEDED | `87522b94c1a`, `67cfadce76e`, rewritten by `97d29a571f8` (#22511), fixed by `2b20cac86fe` (#24165) |
| leverage in `fetchMarkets` | 🔴 STALE (gap partly open, approach unsalvageable) | swap side landed `42bc15fbd5e` (#22180); `margin` flag landed `9c11f72aca4` (#26350); spot `limits.leverage` still undefined |

The gate is 🔴 on the mechanical fact alone — `js/coinex.js` is gone, so there is nothing to merge into. No fault is being found with the code as written for its time; the transfer half was good enough that it shipped, and the leverage idea was right. It has simply been overtaken.

Recommendation: close this PR. If the spot `limits.leverage` gap still matters to you, a new branch on `ts/src/coinex.ts` is a far smaller job than rescuing this one.

**Comment by carlotestor** [comment]
Closing as part of a backlog sweep of conflicting pull requests.

This PR cannot merge (it conflicts with `master`), and the review above found the change is either already on master or no longer applies to the current codebase. The full reasoning, with the specific commits, is here: https://github.com/ccxt/ccxt/pull/12702#pullrequestreview-5192338945

Thank you for the contribution — this is not a judgement on the quality of the work, and if you think the assessment is wrong please say so and it can be reopened.
